### ISSUE TITLE
Improved Full Text Search

### ISSUE THREAD

**Issue by thombruce** [thombruce/notes#11]
At present we can't search the Database by text and order by date at the same time. orderBy and where simply will not work together.

A potential solution looks a bit like this:

```js
this.version(1).stores({
  notes: '$$id,doc,text,createdAt,updatedAt,*textWords,[textWords+updatedAt]'
})
```

In theory, anyway, [textWords+updatedAt] would create a combined index, but I do not know if this is possible with the array-type index that is *textWords.

If it works, it would allow me to do the below:

```js
db.notes
  .where('[textWords+updatedAt]')
  .startsWithAnyOfIgnoreCase(words)
  .offset(offset)
  .limit(limit)
  .reverse()
```

It doesn't actually use the orderBy function (incompatible with where) but it does use the joined index.

...but I do not know if it would even work with startsWithAnyOfIgnoreCase Examples typically use .between(), where the second value uses the min and max key from the index... between essentially orders them.

```js
.between(["foo", Dexie.minKey], ["foo", Dexie.maxKey])
```

The above searches for "foo" in the first value, and orders by the second, so it does seem pretty `.between()` specific.

See also: https://dexie.org/docs/Compound-Index#matching-first-part-only
And: https://github.com/dfahlander/Dexie.js/issues/297#issuecomment-492023301

---

For an idea as to how to add full text search for multiple words at once see: https://github.com/dfahlander/Dexie.js/issues/281

---

Another option for reintroducing orderBy is to... instead of using `.where()` use `.filter()`

`.filter()` works on the results of the initial query, and is not a part of the query itself... so I don't know how it would function with, say, offset and limit... but... we can cycle through the search until we reach the end of the index... That works as a solution, it's just less performative than using a proper index.

**Comment by thombruce** [comment]
The above notes were extracted from a file within the repo. They detail potential solutions to maintaining order while still doing a full search.

For potential solutions concerning search with multiple words (strict - current is permissive allowing OR style contains any word), see: https://github.com/dfahlander/Dexie.js/issues/281

We also might consider installing and utilising Lunr ? There are references to this on the Dexie site. It might be a better indexing solution than our current, crude `textWords` solution which simply separates `text` into an array of words to be matched against.

**Comment by thombruce** [comment]
Multi-word full text search implemented by 939400cff4e3126a2ee183a51a5bc8584ea418fd

This does not feature orderBy or scoring. But note: scoring and orderBy would contradict each other. It's an either or thing.

I believe scoring can be implemented using Lunr.js.

And since the [Lunr implementation](https://flems.io/#0=N4Igxg9gdgzhA2BTEAucD4EMAONEBMQAaEAMwEskZUBtUKTAW2TQDoArakyKAF0T6p0sXgAJMvfo2xiAvKIAiExKygQA7gAoAlAB0oPGGIBOiRuSj5Ex0fOABffYbHwIAc1ujDCFa7esAIwt8TW8kPSgnaCNRYMQAD094AFcoY00dWwA+UQcIqJFRN0ReABUIAGsBAGVeUyZPfni5LP1RWMsE1mxybER4CxVjVM029tEUtNZeSoFyAC9rTSbeCPb8g2ixPExjMAALT018AKJRAEdk6wBPbWzRE9YwXfwYMdZ1fetEUZAZqtguhA2lYiEumHgME0lxua1E0wgAEFjMZMNcdABufT6AD0ONEAAl+n1jG9NoUjLs5KJNFgAv0zrxrn07rIcsAxqZzJ1jDQmX0ALpJTD0+BjMbkUjLZmIUQAQlk8gA5G5jBBktglXD2n5aSL+nCxmFEHyZaIAD7m0RK3jkZhKgV60URRyRckxIwQbBHfmIVnso3RHw0aV9C1Wm12xBa0QAamtAFFLA7NFy4rzfQKXdioAA3CDkfA0zAwa4GGn+3KiMZ1a65MbtSnGXiaJWIyRmGTWuPiDvSXhnFVqjVanPjURNltKhSIIxq64WDz4CQikvR7XVqDjzDqTDkMQz+LkFRWJD8Vv8Iyjt3jz3YHQNze33hU1sAYXqtqgS5XATX1-HR9nAeAJPCgRB1EUBJj1+S9eCBDdHlzawYHIaBNAARhBT1TChYBnmMV4UGrEAADIGGYIgViIAAqf4BDeEB7A3Hc9zER4vQEB8b3GO9uMfSdWwUDQoFcTB8EXLwXhgADxiArZRFw5J4F4GBPFY-dRHAyCAAU1XMPBU1nBBkPuDkt3HUR1GCDRWEYXg3E4aBnngeA-zACpPAI14zMfcdcJMn5gBA4iTjObyYGIiLRGYsdLPaU8SllazLFs+zHLgAwITczAPL8mK4vHAB5AJ2EQMBeFYEtULcKBRgs+L8AgMBkmYPhAggfBriq7A+ksN99koEJ8vHJqWrayqwE-RAEyQCbWxgKbel4WT4u0IgRvrBr4onPZRGIpV9kkXAUDxdKnKgJ4IEYHELpxRFXLfaSOEyzUNu2+LoDEotiM0YAX2MYpeGY3yPp2gGge6XYBF4AA5TqhjMCBkIGobll2IGN3i+x3p2mLccsjdYu2gSZnvDdSdfJVSlRWBSAgYxuQ8CBSvKsQZnEFE0Vk+TCmi+QSrKirWFzCErihJSVJgJ5pI2Z8vX47bBKVABJWBrC-H8X2vR8NPYgIZcI6WAmUipEXwEIIrsnAMnM+KKMQAnxxWR8QbZGk7csh2nfGaj8vo2BiKB8oAVqepGGWBJVmt+8-rnN2cjnIntBTrESaVsnuMAjOqeqRBdgOST6ZsIEypgCo9yBVan3GXU9YnfO9n2Y5TmtMuK-ILU5d4zPu8bAGpzzgvBu-URi5IyBUn4Ywq77iZ3E0eudiblvB0nvhrC7wqJ17x9KebVsh6bouGZI9vK5AavHzr3dNOXg5V7b2cO63nj+4Vimc4PpUj8L0fx6BBDEos897bRvmxBuw9H42gxiUV+pMP5jFdO0Z4vAH5+k9igoMSBQQogZpoP0SCxgUAYK5OsmCd4KyHOqTUc9dTTmgIgOUslXTMR0MQEAeAkAVTQrAIQGEADMKAMIAAYmIChIAMKAFRqAoDoCAB2QgrBHmQCQZIxh4BCCOrwE6eJUjYAqP4SAN1lHHgAAIACZWAiNYAAFhxBJIwDjoIqE4Bw30QhFrGGWkxIg9AmAsBAJMYwHD1GaLQNo3ROJ9GGKujdYJljWACNYAAVhxMEl67iZSeKWjIcR9ggA) also uses `.equals()`, it should also work using `.between()` and a composite index of updatedAt and searchTokens.

**Comment by thombruce** [comment]
Closing for now with a _good enough_ solution. Advanced search may be considered at a later date.
